torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2659,7 +2659,7 @@ def _free_full_params_and_use_local_shard(params_to_free):
                     try:
                         yield
                     finally:
-                        if offload_to_cpu:
+                        if offload_to_cpu and (not rank0_only or my_rank == 0):
```

Comment:
we might also be able to put the below logic into a helper function, as the only thing that differs is  .to(cpu) or .to(device).